ReplicationOperation should fail gracefully

kingherc opened the pull request:
Problem:
finishAsFailed could be called before, or even asynchronously in the middle of, operations like runPostReplicationActions which try to sync the translog. finishAsFailed immediately triggers the failure of the resultListener which releases the index shard primary operation permit. This means that runPostReplicationActions may try to sync the translog without an operation permit.

Solution:
We refactor the infrastructure of ReplicationOperation regarding pendingActions and the resultListener, by replacing them with a RefCountingListener. This way, if there are failures, the result listener is called once, after all pending actions are done.

For the specific error we got in issue #97183, this means that a call to onNoLongerPrimary (which can happen if we fail to fail a replica shard or mark it as stale) will not immediately release the primary operation permit and the assertion in the translog sync will be honored.

Fixes #97183

kingherc committed:
ReplicationOperation should fail gracefully

Problem:
finishAsFailed could be called asynchronously in the
middle of operations like runPostReplicationActions which try to
sync the translog. finishAsFailed immediately triggers the failure
of the resultListener which releases the index shard primary
operation permit. This means that runPostReplicationActions may
try to sync the translog without an operation permit.

Solution:
We refactor the infrastructure of ReplicationOperation regarding
pendingActions and the resultListener, by replacing them with a
RefCountingListener. This way, if there are async failures, they
are aggregated, and the result listener is called once, after all
mid-way operations are done.

For the specific error we got in issue #97183, this means that
a call to onNoLongerPrimary (which can happen if we fail to fail
a replica shard or mark it as stale) will not immediately release
the primary operation permit and the assertion in the translog sync
will be honored.

Fixes #97183

kingherc committed:
Merge remote-tracking branch 'origin/main' into test-failure/97183-translog-shutdown

kingherc committed:
Merge remote-tracking branch 'origin/main' into test-failure/97183-translog-shutdown

Tim-Brooks commented:
I should be able to dig into this tomorrow.

kingherc committed:
Merge remote-tracking branch 'origin/main' into test-failure/97183-translog-shutdown

Tim-Brooks commented on server/src/main/java/org/elasticsearch/action/support/replication/ReplicationOperation.java:
```diff
@@ -106,28 +113,33 @@ public ReplicationOperation(
-        primary.perform(request, ActionListener.wrap(this::handlePrimaryResult, this::finishAsFailed));
+            totalShards.incrementAndGet();
+            var primaryCoordinationPendingActionListener = pendingActionsListener.acquire(); // triggered when we finish all coordination
+            primary.perform(request, ActionListener.wrap(primaryResult -> {
```
Maybe

```java
            totalShards.incrementAndGet();
            var primaryCoordinationPendingActionListener = pendingActionsListener.acquire(); // triggered when we finish all coordination
            primary.perform(request, primaryCoordinationPendingActionListener.delegateFailureAndWrap(this::handlePrimaryResult));
        }
    }

    private void handlePrimaryResult(ActionListener<Void> primaryCoordinationPendingActionListener, final PrimaryResultT primaryResult) {
        this.primaryResult = primaryResult;
        final ReplicaRequest replicaRequest = primaryResult.replicaRequest();
        if (replicaRequest != null) {
            if (logger.isTraceEnabled()) {
                logger.trace("[{}] op [{}] completed on primary for request [{}]", primary.routingEntry().shardId(), opType, request);
            }
            final ReplicationGroup replicationGroup = primary.getReplicationGroup();
```

Tim-Brooks commented on server/src/main/java/org/elasticsearch/action/support/replication/ReplicationOperation.java:
```diff
@@ -193,20 +205,28 @@ public void onFailure(Exception e) {
-                ActionListener.wrap(r -> decPendingAndFinishIfNeeded(), ReplicationOperation.this::onNoLongerPrimary)
+                ActionListener.wrap(
+                    r -> staleCopyPendingActionListener.onResponse(null),
+                    e -> onNoLongerPrimary(e, staleCopyPendingActionListener)
```
Maybe

```
staleCopyPendingActionListener.delegateResponse((listener, e) -> onNoLongerPrimary(e, listener))
```

Tim-Brooks commented on server/src/main/java/org/elasticsearch/action/support/replication/ReplicationOperation.java:
```diff
@@ -282,7 +307,10 @@ public void onFailure(Exception replicaException) {
                     message,
                     replicaException,
-                    ActionListener.wrap(r -> decPendingAndFinishIfNeeded(), ReplicationOperation.this::onNoLongerPrimary)
+                    ActionListener.wrap(
```
Maybe 

```java
replicationPendingActionListener.delegateResponse(
                        (listener, e) -> onNoLongerPrimary(e, replicationPendingActionListener)
                    )
```

Tim-Brooks commented on server/src/main/java/org/elasticsearch/action/support/replication/ReplicationOperation.java:
```diff
@@ -369,13 +397,13 @@ public void onAfter() {
     }
 
-    private void onNoLongerPrimary(Exception failure) {
+    private void onNoLongerPrimary(Exception failure, ActionListener<Void> pendingActionlistener) {
```
Maybe just named titled `listener` to more differentiate between this being local scoped and the field name.

Tim-Brooks reviewed (approved):
LGTM. All the comments are optional. Just things around listener methods.

kingherc committed:
Merge remote-tracking branch 'origin/main' into test-failure/97183-translog-shutdown

kingherc committed:
PR feedback and a bit safer listeners

Such as moving RefCountingListener to try with resources, and
adding ActionListener.run in a couple of places.

kingherc committed:
Merge remote-tracking branch 'origin/main' into test-failure/97183-translog-shutdown

kingherc commented:
Thanks @Tim-Brooks ! Incorporated your feedback, and also made some nit changes to ensure no exception escapes through (by making use of try-with-resources for the ActionListenerRef and some ActionListener.runs in a couple of places). Feel free to give another quick look if you'd like.

Tim-Brooks commented:
> Feel free to give another quick look if you'd like.

LGTM
